File: services/api/app/schemas/questions.py
"""题目相关 Pydantic schemas"""
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class QuestionResponse(BaseModel):
    """题目响应"""

    id: int
    subject: str
    # 数据源 JSON 中 knowledge_points/options 可能是 list[str] 或 list[dict]
    # 用 list[Any] 兼容两种结构, 避免数据导入后 schema 校验失败
    knowledge_points: Optional[list[Any]] = None
    type: str
    difficulty: int
    title: Optional[str] = None
    content: str
    options: Optional[list[Any]] = None
    answer: Optional[str] = None  # 列表或详情页隐藏，练习页不返回
    solution: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class QuestionListResponse(BaseModel):
    """题目列表响应"""

    items: list[QuestionResponse]
    total: int
    page: int = 1
    page_size: int = 20


class QuestionAttemptRequest(BaseModel):
    """提交答案请求"""

    user_answer: str = Field(..., min_length=1)
    duration_ms: Optional[int] = Field(None, ge=0, description="作答耗时（毫秒）")

    model_config = {"extra": "forbid"}


class QuestionAttemptResponse(BaseModel):
    """提交答案响应"""

    correct: bool
    correct_answer: Optional[str] = None
    solution: Optional[str] = None
    xp_gained: int = 0
    mastery_update: Optional[dict[str, Any]] = None


class RecommendRequest(BaseModel):
    """推荐题目请求参数"""

    subject: Optional[str] = None
    count: int = Field(default=10, ge=1, le=50)
    exclude_ids: Optional[list[int]] = None


class RecommendResponse(BaseModel):
    """推荐题目响应"""

    questions: list[QuestionResponse]
    recommendation_reason: Optional[str] = None

File: apps/student-web/src/api/questions.ts
import client from './client';

/** 题目条目 */
export interface QuestionItem {
  id: string;
  subject: string;
  type: string;
  difficulty: string;
  title: string;
  /** 真正的题目正文/题干（后端 content 字段） */
  content?: string;
  options?: string[] | Record<string, string>;
  answer?: string;
  /** 后端字段为 solution，前端兼容用 analysis */
  analysis?: string;
  solution?: string;
  knowledge_points?: string[];
}

/** 题目列表响应 */
export interface QuestionListResponse {
  items: QuestionItem[];
  total: number;
  page?: number;
  page_size?: number;
}

/** 题目详情 */
export interface QuestionDetail {
  id: string;
  subject: string;
  type: string;
  difficulty: string;
  title: string;
  /** 真正的题目正文/题干（后端 content 字段） */
  content?: string;
  options: string[] | Record<string, string>;
  answer: string;
  /** 后端字段为 solution，前端兼容用 analysis */
  analysis?: string;
  solution?: string;
  knowledge_points: string[];
}

/** 答题结果（对齐后端 QuestionAttemptResponse） */
export interface AttemptResult {
  correct: boolean;
  correct_answer: string | null;
  /** 后端字段名为 solution，非 analysis */
  solution: string | null;
  xp_gained?: number;
}

/** 获取题目列表参数 */
export interface GetQuestionsParams {
  subject?: string;
  type?: string;
  difficulty?: string;
  category?: string;
  kp_id?: string;
  page?: number;
}

/**
 * 题目 API 封装层。
 */
export const questionsApi = {
  /**
   * 获取题目列表（分页 + 筛选）
   * GET /api/v1/questions
   * 后端返回 solution 字段，前端兼容映射到 analysis
   */
  async getQuestions(params: GetQuestionsParams = {}): Promise<QuestionListResponse> {
    const res = await client.get('/api/v1/questions', {
      params: {
        subject: params.subject,
        type: params.type,
        difficulty: params.difficulty,
        category: params.category,
        kp_id: params.kp_id,
        page: params.page || 1,
      },
    });
    // 字段映射：solution → analysis（前端页面用 analysis）
    const data = res.data;
    if (data && Array.isArray(data.items)) {
      data.items.forEach((item: QuestionItem) => {
        if (item.solution !== undefined && item.analysis === undefined) {
          item.analysis = item.solution;
        }
      });
    }
    return data;
  },

  /**
   * 获取题目详情
   * GET /api/v1/questions/{id}
   * 后端返回 solution 字段，前端兼容映射到 analysis
   */
  async getQuestionDetail(id: string): Promise<QuestionDetail> {
    const res = await client.get(`/api/v1/questions/${id}`);
    const data = res.data;
    if (data && data.solution !== undefined && data.analysis === undefined) {
      data.analysis = data.solution;
    }
    return data;
  },

  /**
   * 提交答题（判分）
   * POST /api/v1/questions/{id}/attempt
   *
   * 后端 QuestionAttemptRequest 要求字段 duration_ms（毫秒），且禁止多余字段。
   */
  async submitAttempt(
    id: string,
    userAnswer: string,
    timeSpent: number
  ): Promise<AttemptResult> {
    const res = await client.post(`/api/v1/questions/${id}/attempt`, {
      user_answer: userAnswer,
      duration_ms: timeSpent,
    });
    return res.data;
  },
};
